src/twister2/quarantine_plugin.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+import pytest
+import logging
+
+from pathlib import Path
+from yaml import safe_load
+from marshmallow import Schema, fields, ValidationError
+from dataclasses import dataclass, field
+
+from twister2.report.helper import get_test_name, get_item_platform
+from twister2.exceptions import TwisterConfigurationException
+
+logger = logging.getLogger(__name__)
+
+
+class QuarantinePlugin:
+    """Handle tests under quarantine"""
+
+    def __init__(self, config: pytest.Config) -> None:
+        """
+        :param config: pytest configuration
+        """
+        self.quarantine_verify = 'quarantine' in config.getoption('-m')
+        self.quarantine = QuarantineData()
+        for quarantine_file in config.getoption('--quarantine-list'):
+            self.quarantine.extend(QuarantineData.load_data_from_yaml(quarantine_file))
+
+    def pytest_collection_modifyitems(
+        self, session: pytest.Session, config: pytest.Config, items: list[pytest.Item]
+    ):
+        for item in items:
+            if qelem := self.quarantine.get_matched_quarantine(item):
+                item.add_marker(pytest.mark.quarantine(qelem.comment))
+                # do not mark tests to be skipped if want to verify quarantined tests
+                if not self.quarantine_verify:
+                    item.add_marker(pytest.mark.skip('under quarantine'))
+                    logger.debug("Skipped test %s - quarantine reason: %s" % (get_test_name(item), qelem.comment))
+
+
+@dataclass
+class QuarantineElement:
+    scenarios: list[str] = field(default_factory=list)
+    platforms: list[str] = field(default_factory=list)
+    architectures: list[str] = field(default_factory=list)
+    comment: str = 'under quarantine'
+
+    def __post_init__(self):
+        if 'all' in self.scenarios:
+            self.scenarios = []
+        if 'all' in self.platforms:
+            self.platforms = []
+        if 'all' in self.architectures:
+            self.architectures = []
+        if not any([self.scenarios, self.platforms, self.architectures]):
+            raise TwisterConfigurationException(
+                "At least one of filters ('scenarios', 'platforms', 'architectures') must be specified")
+
+
+@dataclass
+class QuarantineData:
+    qlist: list[QuarantineElement] = field(default_factory=list)
+
+    def __post_init__(self):
```

Comment:
This will not work if you pass list of `QuarantineElement` instead of list of dicts like so:
```python
qd = QuarantineData([QuarantineElement])
```
Which is what I would do seeing this class. 

I suggest two solution:
- check in `__post_init__` if self.qlist is a list of dict
- add new class method method e.g. `QuarantineData.from_row_data()` which return QuarantuneData instance

Reply:
updated __post_init__ method. Please check it now. Thanks!